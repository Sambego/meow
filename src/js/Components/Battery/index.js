import {h, render, Component} from 'preact';
import classnames from 'classnames';
import {BubbleSlide, BubbleCollection, Bubble, Icon} from '../../Components';
import styles from './battery.scss';

export default class Battery extends Component {
    constructor(...props) {
        super(...props);

        this.state = {
            level: 1,
        };

        if ('getBattery' in navigator) {
            ::navigator.getBattery().then(battery => {
                this.setState({
                    battery,
                    charging: battery.charging,
                    level: battery.level,
                });

                this.state.battery.addEventListener('levelchange', this.levelListener);
                this.state.battery.addEventListener('chargingchange', this.chargingListener);
            });
        }
    }

    levelListener = () => this.setBatteryLevel();

    chargingListener = () => this.setChargingState();

    setBatteryLevel() {
        this.setState({
            level: this.state.battery.level,
        });
    }

    setChargingState() {
        this.setState({
            charging: this.state.battery.charging,
        });
    }

    componentWillUnmount() {
        this.state.battery.removeEventListener('levelchange', this.levelListener);
        this.state.battery.removeEventListener('chargingchange', this.chargingListener);
    }

    render() {
        const classes = classnames(styles['battery__inner'], {
            [styles['battery__inner--low']]: this.state.level < 0.3,
        });

        return (
            <div className={styles.battery}>
                <span className={styles['battery__text']}>
                    {this.state.charging && <Icon name="lightning" size="text" className={styles['battery__lightning']}/>} {`${this.state.level * 100}%`}</span>
                <div className={classes} style={`width: ${this.state.level * 100}%;`}></div>
            </div>
        );
    }
};
