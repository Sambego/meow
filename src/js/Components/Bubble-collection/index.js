import {h, render} from 'preact';
import {element} from 'proptypes';
import styles from  './Bubble-collection.scss';

const BubbleCollection = ({children}) => {
    return (
        <div className={styles['bubble-collection']}>
            <div className={styles['bubble-collection__inner']}>
                {children}
            </div>
        </div>
    );
};

BubbleCollection.propTypes = {
    children: element,
};

export default BubbleCollection;
