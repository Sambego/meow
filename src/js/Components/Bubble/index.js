import {h, render} from 'preact';
import {element, string, boolean} from 'proptypes';
import classnames from 'classnames';
import styles from './bubble.scss';

const Bubble = ({children, me, full}) => {
    const classes = classnames(styles.bubble, {
        [styles['bubble--me']]: me,
        [styles['bubble--full']]: full,
    });

    return (
        <div className={classes}>
            {children}
        </div>
    );
};

Bubble.propTypes = {
    children: element,
    full: boolean,
    me: boolean,
};

Bubble.defaultProps = {
    me: false,
};

export default Bubble;
