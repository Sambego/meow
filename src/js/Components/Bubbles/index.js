import {h, render, Component} from 'preact';
import {element, func} from 'proptypes';
import keycode from 'keycode';
import styles from './bubbles.scss';

export default class Bubbles extends Component {
    static propTypes = {
        children: element,
        onNextBubble: func,
    };

    constructor(...props) {
        super(...props);

        this.state = {
            shownBubbles: 1,
        };

        if (typeof this.props.onNextBubble !== 'undefined') {
            this.props.onNextBubble(this.state.shownBubbles);
        }

        window.addEventListener('keyup', this.nextBubbleInstance);
    }

    renderBubbles() {
        return this.props.children.map((child, index) => {
            if (index <= (this.state.shownBubbles - 1)) {
                return child;
            }

            return null;
        });
    };

    nextBubble(event) {
        if (this.state.shownBubbles <= this.props.children.length && keycode(event) === 'up') {
            this.setState({shownBubbles: (this.state.shownBubbles + 1)});

            if (typeof this.props.onNextBubble !== 'undefined') {
                this.props.onNextBubble(this.state.shownBubbles);
            }
        } else {
            window.removeEventListener('keyup', this.nextBubbleInstance);
        }
    };

    nextBubbleInstance = event => this.nextBubble(event);

    render() {
        return (
            <div className={styles.bubbles}>
                {this.renderBubbles()}
            </div>
        );
    }
};
