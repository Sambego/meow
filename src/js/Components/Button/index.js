import {h, render, Component} from 'preact';
import {string, element, object, func} from 'proptypes';
import classnames from 'classnames';
import {Icon} from '../';
import styles from './button.scss';

const aliases = {};

const Button = ({children, type, onClick, href, icon, className}) => {
    if (aliases.hasOwnProperty(name)) {
        name = aliases[name];
    }

    const props = {
        className: classnames(styles.button, styles[`button--${type}`], styles[className]),
        onClick,
    };

    if (href) {
        props.href = href;
    }

    if (href) {
        return (
            <a {...props}>
                {icon && <Icon name={icon} className={styles['button__icon']} />}
                <span className={styles['button__inner']}>{children}</span>
            </a>
        );
    }

    return (
        <button {...props}>
            {icon && <Icon name={icon} className={styles['button__icon']} />}
            <span className={styles['button__inner']}>{children}</span>
        </button>
    );
};

Button.propTypes = {
    children: element.isRequired,
    className: string,
    icon: string,
    onClick: func,
    href: string,
    style: object,
    type: string,
};

Button.defaultProps = {
    type: 'primary',
};

export default Button;
