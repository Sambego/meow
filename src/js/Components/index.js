export Bubble from './Bubble';
export Bubbles from './Bubbles';
export Button from './Button';
export Cat from './Cat';
export Code from './Code';
export Container from './Container';
export Footer from './Footer';
export Icon from './Icon';
export Slide from './Slide';
export SlideNav from './SlideNav';
