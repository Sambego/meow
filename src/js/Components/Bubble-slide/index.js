import {h, render, Component} from 'preact';
import linkRef from 'linkref';
import {Slide, Bubbles, Icon} from '../../Components';
import {Keyboard} from '../../Services';
import styles from './bubble-slide.scss';

export default class BubbleSlide extends Component {
    goToPreviousPage() {
        this.refs.slide.previous();
    }

    goToNextPage() {
        this.refs.slide.next();
    }

    getPreviousAction() {
        if (!this.refs.bubbles.firstBubbleIsShown()) {
            return this.refs.bubbles.previousBubble();
        }

        return this.goToPreviousPage();
    }

    getNextAction() {
        if (!this.refs.bubbles.lastBubbleIsShown()) {
            return this.refs.bubbles.nextBubble();
        }

        return this.goToNextPage();
    }

    resetBubbles() {
        this.refs.bubbles.reset();
    }

    scroll() {
        this.refs.slide.base.scrollTop = (this.refs.slide.base.scrollHeight - this.refs.slide.base.clientHeight);
    }

    componentWillMount() {
        this.KeyboardLeftListener = Keyboard.on('left', () => this.getPreviousAction());
        this.KeyboardRightListener = Keyboard.on('right', () => this.getNextAction());
        this.KeyboardUpListener = Keyboard.on('page up', () => this.getPreviousAction());
        this.KeyboardDownListener = Keyboard.on('page down', () => this.getNextAction());
    }

    componentWillUnmount() {
        this.resetBubbles();

        Keyboard.off(this.KeyboardLeftListener);
        Keyboard.off(this.KeyboardRightListener);
        Keyboard.off(this.KeyboardUpListener);
        Keyboard.off(this.KeyboardDownListener);
    }

    render() {
        return (
            <Slide previous={this.props.previous} next={this.props.next} ref={linkRef(this, 'slide')}>
                <Bubbles ref={linkRef(this, 'bubbles')} onUpdate={::this.scroll}>
                    {this.props.children}
                </Bubbles>

                <nav className={styles['bubble-nav']}>
                    <button className={styles['bubble-nav__prev']} onClick={::this.getPreviousAction}><Icon name="caret-left" color="#66B7F9" size="small"/></button>
                    <button className={styles['bubble-nav__next']} onClick={::this.getNextAction}><Icon name="caret-right" color="#66B7F9" size="small" /></button>
                </nav>
            </Slide>
        );
    }
};
