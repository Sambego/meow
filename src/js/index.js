import '../scss/styles.scss';

import {h, render} from 'preact';
import Router from 'preact-router';
import {Container, Footer} from './Components';
import {
    StartPage,
    SpeechPage,
    SpeechCodePage,
} from './Pages';

render((
    <div>
        <Container>
            <Router>
                <StartPage path="/" />
                <SpeechPage path="/speech" />
                <SpeechCodePage path="/speech-code" />
            </Router>
        </Container>

        <Footer />
    </div>
), document.body);
