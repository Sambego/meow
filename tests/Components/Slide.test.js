import render from 'preact-render-to-string';
import {h, render as PreactRender} from 'preact';
import {Slide} from '../../src/js/Components';

describe('Icon', () => {
    const content = <p>content</p>;
    const previous = '/previous';
    const next = '/next';

    it('should render correctly', () => {
        const slide = render(
            <Slide previous={previous} next={next}>
                {content}
            </Slide>
        );

        expect(slide).toMatchSnapshot();
    });

    it('should add a previous slide button', () => {
        const slide = PreactRender(
            <Slide previous={previous}>
                {content}
            </Slide>
        );

        expect(slide.querySelector(`a[href="${previous}"]`)).toBeTruthy();
        expect(slide.querySelector(`a[href="${next}"]`)).toBeFalsy();
    });

    it('should add next slide button', () => {
        const slide = PreactRender(
            <Slide next={next}>
                {content}
            </Slide>
        );

        expect(slide.querySelector(`a[href="${previous}"]`)).toBeFalsy();
        expect(slide.querySelector(`a[href="${next}"]`)).toBeTruthy();
    });

    it('should add both previous and next slide buttons', () => {
        const slide = PreactRender(
            <Slide previous={previous} next={next}>
                {content}
            </Slide>
        );

        expect(slide.querySelector(`a[href="${previous}"]`)).toBeTruthy();
        expect(slide.querySelector(`a[href="${next}"]`)).toBeTruthy();
    });
});
