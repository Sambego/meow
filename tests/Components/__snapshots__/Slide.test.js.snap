exports[`Icon should render correctly 1`] = `"<section><a class=\"slide-nav--left\" href=\"/previous\"><span class></span></a><a class=\"slide-nav--right\" href=\"/next\"><span class></span></a><p>content</p></section>"`;
